### PR TITLE
Spacepods

### PR DESCRIPTION
**Pull request by Gandalf2k15**
## About The Pull Request

After two long years, we are finally getting spaceballs. The overmap never came.

### Physics component
This now adds a component that processes tile based pixel physics. You can apply this to anything, give it vector thrust and angular momentum. It processes impacts, drag etc all itself.

### Spacepods
Adds an entirely new mode of transport in space, the spacepod.

These spacepods use physics to calculate movement which is not tile-based like most other spacepods. They have collision physics and damage physics. 

Spacepods have a dedicated GUI and many action buttons. They are capable of having weapons and equipment mounted.

Currently, there is a limited list of equipment, however, it's very easy to add more and is not just limited to weapons. 

You can find spacepods in space(soon TM) and you can build them after researching them, or be the spacepod pilot.

Gallery:
A basic sec spacepod
image

Spaceballs:
https://user-images.githubusercontent.com/9026500/230704944-9ac65b47-3055-4dab-b3c3-9731689b41a2.mp4

Idiot crashes a spacepod:
https://user-images.githubusercontent.com/9026500/230704959-cb3244be-f832-44b7-8f65-2e5e903c0b5b.mp4

Spacepod damage:
https://user-images.githubusercontent.com/9026500/230705111-9236ec99-19a0-43e0-ab4b-e5826ce823a5.mp4

### Spacepod Pilot
Due to the fact we now have spacepods that can be found and made throughout space, it only makes sense to have at least one job dedicated to it. Thus you get the spacepod pilot.

The spacepod pilot is considered to be part of security(much like all other servers). They have a small hangar bay with a partially completed spacepod. This pod will come with a simple disabler unit until science gets upgrades.

They get a unique uniform.

image

## Credits
Some physics code from https://github.com/frosty-dev/white
Spacepod sprites go to arctasia

## TO-DO
todo
- [x] add button to stop rotation
- [ ] add siren equipment
- [ ] add missile weapon
- [ ] add machine gun
- [ ] add burst fire plasma cutter
- [ ] fix weapon origin bug
- [ ] add hud alert for low battery
- [ ] add check to make sure there are not more people than can handle

- [x] assoc list testing:
- [x] pod destruction
- [x] pod deletion
- [x] pod removal of seats w/ someone in
- [x] pod enter
- [x] pod exit
- [ ] pod drag someone in

- [ ] add gps system

- [x] left weapon does not have correct offset
- [x] add target lock hud warning
- [ ] add weapon hud
- [ ] add speed/heading hud

- [ ] refactor mouse move

## Changelog

:cl:
add: Added pilotable spacepods to the game.
add: Added a new job: The spacepod pilot!
/:cl:

### CODE REVIEW

**Comment by CandleJaxx** [comment]
very poggers

**Comment by Gandalf2k15** [comment]
reopening because I stopped work on these for the overmap lol, so much for that.

**Comment by Iamgoofball** [comment]
aight but how are you handling the multi-tile object problem

File: code/game/machinery/porta_turret/portable_turret.dm [hunk 1]
```diff
@@ -490,6 +490,18 @@ DEFINE_BITFIELD(turret_flags, list(
 					if(assess_perp(occupant) >= 4)
 						targets += mech
 
+	// SKYRAT EDIT ADDITION
+	for(var/obj/spacepod/iterating_spacepod as anything in GLOB.spacepods_list)
```
Comment (Zonespace27) [inline code/game/machinery/porta_turret/portable_turret.dm:494]: i hate that this is the way it has to be done.

File: code/__DEFINES/~skyrat_defines/signals.dm [hunk 2]
```diff
@@ -64,6 +64,9 @@
 //when someone pulls back their fishing rod
 #define COMSIG_FINISH_FISHING "finish_fishing"
 
+/// /client/MouseMove(object, location, control, params)
+#define COMSIG_MOB_CLIENT_MOUSE_MOVE "client_mob_mouse_move"
```
Comment (Zonespace27) [inline code/__DEFINES/~skyrat_defines/signals.dm:68]: MOB_CLIENT_MOUSE_MOVE vs client_mob_mouse_move

File: modular_skyrat/modules/spacepods/code/spacepods/procs/spacepod_rider_procs.dm [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+ *
+ * rider_types can be a list or a singleton
+ */
+/obj/spacepod/proc/to_chat_to_riders(rider_types, message)
+	if(islist(rider_types))
+		for(var/rider_type as anything in rider_types)
+			for(var/occupant as anything in get_all_occupants_by_type(rider_type))
+				to_chat(occupant, message)
+		return
+	for(var/occupant as anything in get_all_occupants_by_type(rider_types))
```
Comment (Zonespace27) [inline modular_skyrat/modules/spacepods/code/spacepods/procs/spacepod_rider_procs.dm]: as anything is redundant here

File: modular_skyrat/modules/spacepods/code/spacepods/procs/spacepod_rider_procs.dm [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+/**
+ * to_chat_to_riders
+ *
+ * Sends a message to a list of rider types, or just a rider type.
+ *
+ * rider_types can be a list or a singleton
+ */
+/obj/spacepod/proc/to_chat_to_riders(rider_types, message)
+	if(islist(rider_types))
+		for(var/rider_type as anything in rider_types)
```
Comment (Zonespace27) [inline modular_skyrat/modules/spacepods/code/spacepods/procs/spacepod_rider_procs.dm]: and here

File: modular_skyrat/modules/spacepods/code/spacepods/procs/spacepod_rider_procs.dm [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+			for(var/occupant as anything in get_all_occupants_by_type(rider_type))
```
Comment (Zonespace27) [inline modular_skyrat/modules/spacepods/code/spacepods/procs/spacepod_rider_procs.dm]: and here

File: modular_skyrat/modules/spacepods/code/spacepods/procs/spacepod_rider_procs.dm [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+/**
+ * exit pod
+ *
+ * Makes the user exit the pod.
+ */
+/obj/spacepod/proc/exit_pod(mob/user)
+	if(HAS_TRAIT(user, TRAIT_RESTRAINED))
+		to_chat(user, span_notice("You attempt to stumble out of [src]. This will take two minutes."))
+		to_chat_to_riders(SPACEPOD_RIDER_TYPE_PILOT, span_warning("[user] is trying to escape [src]."))
+		if(!do_after(user, 1200, target = src))
```
Comment (Zonespace27) [inline modular_skyrat/modules/spacepods/code/spacepods/procs/spacepod_rider_procs.dm]: MINUTES define please

File: code/__DEFINES/~skyrat_defines/spacepod_helpers.dm [hunk 7]
```diff
@@ -0,0 +1,8 @@
+#define isspacepod(A) (istype(A, /obj/spacepod))
+#define isspacepodequipment(A) (istype(A, /obj/item/spacepod_equipment))
+
+/// Sends a signal when the client detects mouse movement
+/client/MouseMove(atom/object, atom/location, control, params)
+	. = ..()
+	if(mob)
+		SEND_SIGNAL(mob, COMSIG_MOB_CLIENT_MOUSE_MOVE, object, location, control, params)
```
Comment (Zonespace27) [inline code/__DEFINES/~skyrat_defines/spacepod_helpers.dm:5]: I am incredibly, incredibly sus of this code
to quote MSO
image

image



Bare minimum, I want this thoroughly profiled on live before merge

File: code/__DEFINES/~skyrat_defines/signals.dm [hunk 8]
```diff
@@ -82,3 +85,39 @@
...
+/// Sent when a physics component processes a physics based bump. (bump_velocity, bumped_atom)
+#define COMSIG_PHYSICS_PROCESSED_BUMP "physics_processed_bump"
+
+/// Sent to check if the physics component can apply thrust to the atom. (source, total_x, total_y, desired_thrust_dir, delta_time)
+#define COMSIG_PHYSICS_THRUST_CHECK "physics_thrust_check"
+	/// Return bitflag for enabled thrust.
+	#define COMPONENT_PHYSICS_THRUST (1<<0)
+
+/// Sent to check if the physics component can autostabalise. (source)
+#define COMSIG_PHYSICS_AUTOSTABALISE_CHECK "physics_autostabalise_check"
```
Comment (Zonespace27) [inline code/__DEFINES/~skyrat_defines/signals.dm]: that's not how stabilize is spelled

File: code/__DEFINES/~skyrat_defines/spacepod_defines.dm [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+// Slots
+#define SPACEPOD_SLOT_CARGO "cargo"
+#define SPACEPOD_SLOT_MISC "misc"
+#define SPACEPOD_SLOT_WEAPON "weapon"
+#define SPACEPOD_SLOT_LOCK "lock"
+#define SPACEPOD_SLOT_LIGHT "light"
+#define SPACEPOD_SLOT_THRUSTER "thruster"
+
+#define SPACEPOD_LAYER 3.9
```
Comment (Zonespace27) [inline code/__DEFINES/~skyrat_defines/spacepod_defines.dm]: this should be in the layer file (if we have one)

File: code/__DEFINES/~skyrat_defines/spacepod_helpers.dm [hunk 10]
```diff
@@ -0,0 +1,8 @@
+#define isspacepod(A) (istype(A, /obj/spacepod))
+#define isspacepodequipment(A) (istype(A, /obj/item/spacepod_equipment))
```
Comment (Zonespace27) [inline code/__DEFINES/~skyrat_defines/spacepod_helpers.dm]: these should be in the ishelpers file

**Review by Zonespace27 — CHANGES_REQUESTED** [review]
haphazard review of a few files

File: code/__DEFINES/~skyrat_defines/spacepod_helpers.dm (hunk 7 above)
Comment (Iamgoofball) [reply code/__DEFINES/~skyrat_defines/spacepod_helpers.dm:5]: as a note, there is a cheap way to do this

look at how the Kinesis module works for MODSuits(2021) (by fikou)

Comment (Gandalf2k15) [reply code/__DEFINES/~skyrat_defines/spacepod_helpers.dm:5]: Its on the to do

**Comment by AffectedArc07** [comment]
The physics stress test means nothing on a local server when the real one is gonna have 160 people and 40% TD already.

`animate()` calls scale with player count and will drastically nuke perf. You have been warned.

**Comment by Gandalf2k15** [comment]
> 

This has been run on a fairly populated server for some time, I doubt it will gave much impact if any considering even then there won't be many instances at all.

**Comment by Zonespace27** [comment]
__make sure the licensing is okay with the goon sprites__

**Comment by Gandalf2k15** [comment]
> **make sure the licensing is okay with the goon sprites**

resolved, I have my own now

**Comment by YakumoChen** [comment]
PodSec feels like an under developed role for multi hour rounds where sec is generally expected to stay on the z level. They can't even really hunt carp with the default gear.

I could see it getting use on an antag heavy server like goon or vg, but what does this job have for content when nobody is fucking around space?

File: modular_skyrat/modules/spacepods/code/spacepods/spacepod_prebuilt.dm [hunk 11]
```diff
@@ -0,0 +1,148 @@
+/obj/spacepod/prebuilt
```
Comment (Iamgoofball) [inline modular_skyrat/modules/spacepods/code/spacepods/spacepod_prebuilt.dm:1]: alright, im gonna say the meme:

why isn't this a vehicle?

Comment (Gandalf2k15) [reply modular_skyrat/modules/spacepods/code/spacepods/spacepod_prebuilt.dm:1]: i am so angry

**Comment by Gandalf2k15** [comment]
> PodSec feels like an under developed role for multi hour rounds where sec is generally expected to stay on the z level. They can't even really hunt carp with the default gear.
> 
> I could see it getting use on an antag heavy server like goon or vg, but what does this job have for content when nobody is fucking around space?

For this, they won't be limited and there will be people using other pods. Sec not having a way to combat people in ships would be bad news, and they double as a pod mechanic.

**Comment by YakumoChen** [comment]
> > PodSec feels like an under developed role for multi hour rounds where sec is generally expected to stay on the z level. They can't even really hunt carp with the default gear.
> > I could see it getting use on an antag heavy server like goon or vg, but what does this job have for content when nobody is fucking around space?
> 
> For this, they won't be limited and there will be people using other pods. Sec not having a way to combat people in ships would be bad news, and they double as a pod mechanic.

So they're just Sec+?
Why make it its own role, instead of just letting officers volunteer for it like with sec medic, or like how science department works as a whole?

**Comment by Gandalf2k15** [comment]
still not fucking dead

**Pull request closed without merging**